src/internal/connector/exchange/api/shared.go

```diff
@@ -56,6 +56,62 @@ func toValues[T any](a any) ([]getIDAndAddtler, error) {
 	return r, nil
 }
 
+func getAddedAndRemovedItemIDs(
+	ctx context.Context,
+	service graph.Servicer,
+	user, directoryID, oldDelta string,
+	pagerGetter func(context.Context, graph.Servicer, string, string, bool) (itemPager, error),
+	deltaPagerGetter func(context.Context, graph.Servicer, string, string, string, bool) (itemPager, error),
+	immutableIDs bool,
+) ([]string, []string, DeltaUpdate, error) {
+	var (
+		deltaURL string
+		pgr      itemPager
+		err      error
+	)
+
+	pgr, err = deltaPagerGetter(ctx, service, user, directoryID, oldDelta, immutableIDs)
+	if err != nil {
+		return nil, nil, DeltaUpdate{}, graph.Wrap(ctx, err, "creating delta pager")
+	}
+
+	added, removed, deltaURL, err := getItemsAddedAndRemovedFromContainer(ctx, pgr)
+	// note: happy path, not the error condition
+	if err == nil {
+		return added, removed, DeltaUpdate{deltaURL, len(oldDelta) != 0}, err
+	}
+
+	// return error if invalid delta error or if we did a non-delta fetch
+	if !graph.IsErrInvalidDelta(err) && !graph.IsErrQuotaExceeded(err) {
```

Comment:
@ryanfkeepers I'm switching based on the error instead of trying to see if we fetch via delta at the start as that requires an extra permissions, plus we have validated this behaviour.

Reply:
I have to disagree with that choice.  The core issue is that this is still _reacting_ to an over-quota issue instead of behaving consistently throughout the process runtime.

The overwhelming majority of over-quota cases are issues of user account state, not of folder state.  This approach treats it as folder state.  That means every single folder we enumerate, account-wide, will have to fail, catch the failure, change behavior, and then correct course.  And since the most likely case for over-quota issues is that the user's mailbox will be entirely over quota for the duration of the backup, or not, as a binary condition, this means we'll end up in this fallback for every folder, or none of them.

The purpose of discovering quota state early on is to avoid the per-folder fallback.  How we do the check is flexible; catching the error instead of asking for another permission does solve this case (though I can imagine good reasons to do the latter).  The method isn't as important as the location.  The location we want is at the start of the backup, before any enumeration.  I'm thinking [here where we checkServiceEnabled](https://github.com/alcionai/corso/blob/main/src/internal/connector/data_collections.go#L59), expanding that logic a bit to transform or pass down userInfo.

By generating a high-level flag and passing it to lower layers, we can make clear assumptions about system behavior: _all_ enumeration will avoid deltas, _all_ of the metadata will get dropped in favor of a full backup.  It minimizes the cognitive load about how parts of the system should behave, and when, ensure we have the flag available for all cases that rely on the identification, and increases our capacity to easily test and maintain the code.  The more conditional fallbacks we have, especially those that are dependent on broad account state, the more we encounter questions like "folder A did a fallback, but folder B didn't; is that expected or a bug?" and "what corner cases arise for a backup if we arbitrarily fallback on N of M containers?"

For the most part, the changes you've made here are great and provide stepping stones to handle the behavior as outlined above.  It's just the matter of pinning down how we establish the pivot between delta and non-delta behaviors.  From here, to continue with iteration, I think additional small branches that can be folded back into this one are the best way to iterate and review.  If needed, I'm happy to jump on a call, even if its late for me (morning for you).

Reply:
Thanks for the detailed response. I think you make very good points. For some reason I forgot the fact that we have multiple deltas in exchange. Let me see how I can rework this.

Reply:
Created https://github.com/alcionai/corso/pull/3232 for this.

Reply:
Looking good so far, thank you for the changes!